### PR TITLE
HW 2 - SOLID & GRASP

### PR DESCRIPTION
**Pull request by asemchenko**

### CODE REVIEW

**Comment by asemchenko** [comment]
@anton-uvarenko it's ready for review

File: internal/util/response.go [hunk 1]
```diff
@@ -5,11 +5,12 @@ import (
 	"net/http"
 )
 
-func RespondJSON(w http.ResponseWriter, status int, payload interface{}) {
-	w.Header().Set("Content-Type", "application/json")
-	w.WriteHeader(status)
-	err := json.NewEncoder(w).Encode(payload)
+func RespondJSON(w *http.ResponseWriter, status int, payload interface{}) {
```
Comment (anton-uvarenko) [inline internal/util/response.go]: http.ResponseWriter is interface, so no need in *

File: main.go [hunk 2]
```diff
@@ -35,116 +22,18 @@ var cronInstance *cron.Cron
...
-	if err != nil {
-		log.Fatal("Error initializing currency service: ", err)
-	}
-
-	subscriptionController := controller.NewSubscriptionController(subscriptionService)
-	rateController := controller.NewRateController(currencyService)
+	subscriptionController := controller.NewSubscriptionController(appCtx.SubscriptionService)
+	rateController := controller.NewRateController(appCtx.CurrencyService)
 
 	r.HandleFunc("/api/rate", rateController.GetRate).Methods("GET")
```
Comment (anton-uvarenko) [inline main.go]: i would move routs set up somewhere else

File: main.go [hunk 3]
```diff
@@ -35,116 +22,18 @@ var cronInstance *cron.Cron
...
 	r.HandleFunc("/api/subscribe", subscriptionController.Subscribe).Methods("POST")
 	r.PathPrefix("/swagger/").Handler(httpSwagger.WrapHandler)
 
-	initCron(emailService, currencyService, subscriptionService)
-
 	log.Println("Server started at :8080")
 	log.Fatal(http.ListenAndServe(":8080", r))
```
Comment (anton-uvarenko) [inline main.go]: add graceful shutdown, cause on ctrl+c your connections won't be closed with appCtx.Close()

File: internal/service/currency_service.go [hunk 4]
```diff
@@ -25,96 +18,51 @@ type CurrencyRate struct {
...
-
 type CurrencyService struct {
-	cacheTime  time.Time
 	repo       *repository.ExchangeRateRepository
-	cacheRate  float64
-	cacheMutex sync.Mutex
+	latestRate *util.InMemoryCache[float64]
 }
 
 func NewCurrencyService(repo *repository.ExchangeRateRepository) *CurrencyService {
```
Comment (anton-uvarenko) [inline internal/service/currency_service.go:26]: repository isn't an interface = harder to unit test, same for other repos

File: internal/controller/subscription.go [hunk 5]
```diff
@@ -34,15 +34,18 @@ func (c *SubscriptionController) Subscribe(w http.ResponseWriter, r *http.Reques
 	}
 
 	err := c.Service.Subscribe(email)
-	if err != nil {
-		if errors.Is(err, exceptions.ErrEmailAlreadySubscribed) {
+	renderSubscriptionResponse(&w, &err)
+}
+
+func renderSubscriptionResponse(w *http.ResponseWriter, subscribeError *error) {
```
Comment (anton-uvarenko) [inline internal/controller/subscription.go]: no need in pointers

File: internal/util/response.go (hunk 1 above)
Comment (asemchenko) [reply internal/util/response.go]: fixed

File: main.go (hunk 3 above)
Comment (asemchenko) [reply main.go]: added

File: internal/controller/subscription.go (hunk 5 above)
Comment (asemchenko) [reply internal/controller/subscription.go]: agree

File: internal/service/currency_service.go (hunk 4 above)
Comment (asemchenko) [reply internal/service/currency_service.go:26]: changed in the next MR

https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-asemchenko/pull/4

File: main.go (hunk 2 above)
Comment (asemchenko) [reply main.go]: moved

**Comment by asemchenko** [comment]
@agritsyk are we ready to merge?

**Comment by agritsyk** [comment]
> @agritsyk are we ready to merge?

Так)

**Pull request merged**